File: python/zugferd_facturx_xrechnung/attach_invoice/attach_invoice.py
import sys, os, ctypes
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

# attach_invoice -- Python (ctypes) port of the VB6 mirror.
# Imports an existing PDF/A-3 invoice, attaches the factur-x.xml e-invoice,
# associates it with the catalog and sets the FacturX Comfort PDF version.

HERE = os.path.dirname(os.path.abspath(__file__))
TF = os.path.join(HERE, "..", "..", "..", "test_files")


def _err(data, code, msg, typ):
    if msg:
        print(msg.decode('latin-1', 'replace'))
    return 0


_cb = L.TErrorProc(_err)  # keep global ref


def main():
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, _cb)
    L.pdfCreateNewPDFA(pdf, b"")

    # We assume the pdf invoice is already a valid PDF/A-3 file.
    L.pdfSetImportFlags(pdf, L.ifImportAsPage | L.ifImportAll)
    L.pdfOpenImportFileA(pdf, os.path.join(TF, "test_invoice.pdf").encode('latin-1'), L.ptOpen, b"")

    L.pdfImportPDFFile(pdf, 1, 1.0, 1.0)

    ef = L.pdfAttachFileA(pdf, os.path.join(TF, "factur-x.xml").encode('latin-1'),
                          b"EN 16931 compliant invoice", 0)
    L.pdfAssociateEmbFile(pdf, L.adCatalog, -1, L.arAlternative, ef)

    # ZUGFeRD 2.1+ and FacturX share the same version constants in PDF.
    L.pdfSetPDFVersion(pdf, L.pvFacturX_Comfort)

    if L.pdfHaveOpenDoc(pdf) != 0:
        outFile = os.path.join(HERE, "out.pdf.")
        if L.pdfOpenOutputFileA(pdf, outFile.encode('latin-1')) == 0:
            L.pdfDeletePDF(pdf)
            return
        if L.pdfCloseFile(pdf) != 0:
            print('PDF file "' + outFile + '" successfully created!')
    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: vbnet/zugferd_facturx_xrechnung/attach_invoice/attach_invoice.vb
' attach_invoice -- VB.NET port of examples\Vb6\...\attach_invoice\attach_invoice.bas
' Imports an existing PDF/A-3 invoice, attaches the factur-x.xml e-invoice,
' associates it with the catalog and sets the FacturX Comfort PDF version.
Imports System
Imports LumasPdfSdk

Module modAttachInvoice
    Private ErrDelegate As TErrorProc

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    ' ParamArray is UInteger, not Integer: the generated bindings type the
    ' flag constants as UInteger (they have the high bit set, e.g.
    ' ifImportAsPage = &H80000000UI), and passing one to an Integer
    ' parameter is BC30439 "Constant expression not representable in
    ' type 'Integer'". The arithmetic below is unchanged, so the value
    ' handed to the engine -- and therefore the output -- is identical.
    Function Fl(ParamArray vals() As UInteger) As UInteger
        Dim r As Long = 0
        For Each v In vals : r = r Or (CLng(v) And &HFFFFFFFFL) : Next
        Return CUInt(r And &HFFFFFFFFL)
    End Function

    Public Function PDFError(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return 0                      ' We try to continue if an error occurs
    End Function

    Sub Main()
        Dim ef As Integer, pdf As IntPtr, outFile As String

        pdf = LumasPdf.pdfNewPDF()
        ErrDelegate = New TErrorProc(AddressOf PDFError)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, ErrDelegate)
        LumasPdf.pdfCreateNewPDFW(pdf, "")              ' The output file is opened later

        ' We assume that the pdf invoice is already a valid PDF/A 3 file in this example.
        LumasPdf.pdfSetImportFlags(pdf, Fl(LumasPdfConsts.ifImportAsPage, LumasPdfConsts.ifImportAll))
        LumasPdf.pdfOpenImportFileW(pdf, "../../../test_files/test_invoice.pdf", LumasPdfConsts.ptOpen, "")

        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0)

        ef = LumasPdf.pdfAttachFileW(pdf, "../../../test_files/factur-x.xml", "EN 16931 compliant invoice", False)
        LumasPdf.pdfAssociateEmbFile(pdf, TAFDestObject.adCatalog, -1, TAFRelationship.arAlternative, CUInt(ef))

        ' Note that ZUGFeRD 2.1 or higher and FacturX is identically defined in PDF. Therefore, both formats
        ' share the same version constants.
        LumasPdf.pdfSetPDFVersion(pdf, LumasPdfConsts.pvFacturX_Comfort)

        ' No fatal error occurred?
        If LumasPdf.pdfHaveOpenDoc(pdf) Then
            ' We write the file into the application directory.
            outFile = AppPath() & "\out.pdf."
            If Not LumasPdf.pdfOpenOutputFileW(pdf, outFile) Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            If LumasPdf.pdfCloseFile(pdf) Then
                Console.WriteLine("PDF file """ & outFile & """ successfully created!")
            End If
        End If
        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module
